### PR TITLE
feat(SDRS): add a new resource sdrs replication pair

### PR DESCRIPTION
**Pull request by deer-hang**
**What this PR does / why we need it**:
add a new resource sdrs replication pair
**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #xxx

**Special notes for your reviewer**:

- This PR depends on : feat(SDRS): add a new resource sdrs protection group #3117.
- Make sure the PR #3117 has been merged before merge this PR.

**Release note**:

```release-note

```

## PR Checklist

* [X] Tests added/passed.
* [X] Documentation updated.
* [X] Schema updated.

## Acceptance Steps Performed

```
make testacc TEST='./huaweicloud/services/acceptance/sdrs' TESTARGS='-run TestAccSDRSReplicationPair_basic'
...
==> Checking that code complies with gofmt requirements... 
TF_ACC=1 go test ./huaweicloud/services/acceptance/sdrs -v -run TestAccSDRSReplicationPair_basic -timeout 360m -parallel 4 
=== RUN   TestAccSDRSReplicationPair_basic 
=== PAUSE TestAccSDRSReplicationPair_basic
=== CONT  TestAccSDRSReplicationPair_basic
--- PASS: TestAccSDRSReplicationPair_basic (139.04s) 
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/sdrs      139.099s
```

### CODE REVIEW

File: docs/resources/sdrs_replication_pair.md [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+}
+
+resource "huaweicloud_sdrs_protectiongroup" "test" {
+  name                     = "test-name"
+  source_availability_zone = data.huaweicloud_availability_zones.test.names[0]
+  target_availability_zone = data.huaweicloud_availability_zones.test.names[1]
+  domain_id                = data.huaweicloud_sdrs_domain.test.id
+  source_vpc_id            = data.huaweicloud_vpc.test.id
+  description              = "test description"
+}
```
Comment (chengxiangdong) [inline docs/resources/sdrs_replication_pair.md]:
```suggestion
variable "source_availability_zone" {}
variable "target_availability_zone" {}
variable "domain_id" {}
variable "vpc_id" {}

resource "huaweicloud_sdrs_protectiongroup" "test" {
  name                     = "test-name"
  source_availability_zone = var.source_availability_zone
  target_availability_zone = var.target_availability_zone
  domain_id                = var.domain_id
  source_vpc_id            = var.vpc_id
  description              = "test description"
}
```

It is better to define parameters as variables to highlight the way resource scripts are written.

File: docs/resources/sdrs_replication_pair.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+resource "huaweicloud_evs_volume" "test" {
+  name              = "test-evs-name"
+  description       = "test volume for sdrs replication pair"
+  availability_zone = data.huaweicloud_availability_zones.test.names[0]
```
Comment (chengxiangdong) [inline docs/resources/sdrs_replication_pair.md]: ditto

File: docs/resources/sdrs_replication_pair.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `region` - (Optional, String, ForceNew) Specifies the region in which to create the resource.
+  If omitted, the provider-level region will be used. Changing this parameter will create a new resource.
+
+* `name` - (Required, String) Specifies the name of a replication pair. The name can contain a maximum of 64 bytes.
```
Comment (chengxiangdong) [inline docs/resources/sdrs_replication_pair.md]:
```suggestion
* `name` - (Required, String) Specifies the name of a replication pair. The name can contain a maximum of 64 characters.
```

File: docs/resources/sdrs_replication_pair.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+  The value can contain only letters (a to z and A to Z), digits (0 to 9), decimal points (.), underscores (_),
```
Comment (chengxiangdong) [inline docs/resources/sdrs_replication_pair.md]:
```suggestion
  The value can contain only letters (a to z and A to Z), digits (0 to 9), periods (.), underscores (_),
```

or 

```suggestion
  The value can contain only letters (a to z and A to Z), digits (0 to 9), dots (.), underscores (_),
```

File: docs/resources/sdrs_replication_pair.md [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+  Changing this parameter will create a new resource.
+
+* `delete_target_volume` - (Optional, Bool, ForceNew) Specifies whether to delete the disaster recovery site disk.
+  The default value is **false**.
+
+  Changing this parameter will create a new resource.
+
+* `description` - (Optional, String, ForceNew) Specifies the description of a replication pair. The value can contain
+  a maximum of 64 bytes and cannot contain the left angle bracket (<) or right angle bracket (>).
```
Comment (chengxiangdong) [inline docs/resources/sdrs_replication_pair.md]:
```suggestion
  a maximum of 64 characters and angle brackets (<) and (>) are not allowed.
```

**Review by chengxiangdong — COMMENTED** [review]
TL;DR
Here are some suggestions for docs.

File: huaweicloud/services/sdrs/resource_huaweicloud_sdrs_replication_pair.go [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+		GroupID:     d.Get("group_id").(string),
+		VolumeID:    d.Get("volume_id").(string),
+	}
+
+	n, err := replications.Create(client, createOpts).ExtractJobResponse()
+	if err != nil {
+		return diag.Errorf("error creating SDRS replication pair: %s", err)
+	}
+
+	createTimeoutSec := int(d.Timeout(schema.TimeoutCreate) / time.Second)
```
Comment (chengxiangdong) [inline huaweicloud/services/sdrs/resource_huaweicloud_sdrs_replication_pair.go]: How about this:

```suggestion
	createTimeoutSec := int(d.Timeout(schema.TimeoutCreate).Seconds())
```
It seems easier to understand.

File: huaweicloud/services/sdrs/resource_huaweicloud_sdrs_replication_pair.go [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+
+func resourceSDRSReplicationPairCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	cfg := meta.(*config.Config)
+	region := cfg.GetRegion(d)
+	client, err := cfg.SdrsV1Client(region)
+	if err != nil {
+		return diag.Errorf("error creating SDRS client: %s", err)
+	}
+	// use for WaitForJobSuccess
+	client.Endpoint = client.ResourceBase
```
Comment (chengxiangdong) [inline huaweicloud/services/sdrs/resource_huaweicloud_sdrs_replication_pair.go:93]: client.Endpoint = client.ResourceBase
不理解这个是用来做什么的，而且在 `WaitForJobSuccess` 又互相转换了一次。
如果确实是特殊用途，建议将这个转换放在 `WaitForJobSuccess` 函数内部处理。

File: huaweicloud/services/sdrs/resource_huaweicloud_sdrs_replication_pair.go [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+			"volume_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"delete_target_volume": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				ForceNew: true,
+			},
```
Comment (chengxiangdong) [inline huaweicloud/services/sdrs/resource_huaweicloud_sdrs_replication_pair.go]:
```suggestion
			"delete_target_volume": {
				Type:     schema.TypeBool,
				Optional: true,
			},
```
Changing this does not create a new resource.

File: huaweicloud/services/sdrs/resource_huaweicloud_sdrs_replication_pair.go [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+	// use for WaitForJobSuccess
+	client.Endpoint = client.ResourceBase
+
+	deleteOpts := replications.DeleteOpts{
+		GroupID:      d.Get("group_id").(string),
+		DeleteVolume: d.Get("delete_target_volume").(bool),
+	}
+	n, err := replications.Delete(client, d.Id(), deleteOpts).ExtractJobResponse()
+	if err != nil {
+		return diag.Errorf("error deleting SDRS replication pair, %s", err)
```
Comment (chengxiangdong) [inline huaweicloud/services/sdrs/resource_huaweicloud_sdrs_replication_pair.go]:
```suggestion
		return common.CheckDeletedDiag(d, err, "error deleting SDRS replication pair")
```

File: huaweicloud/services/acceptance/sdrs/resource_huaweicloud_sdrs_replication_pair_test.go [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+}
+
+resource "huaweicloud_evs_volume" "test" {
+  name              = "%[1]s"
+  description       = "test volume for sdrs replication pair"
+  availability_zone = data.huaweicloud_availability_zones.test.names[0]
+  volume_type       = "SSD"
+  size              = 100
+}
+`, name)
```
Comment (chengxiangdong) [inline huaweicloud/services/acceptance/sdrs/resource_huaweicloud_sdrs_replication_pair_test.go]:
```suggestion
	return fmt.Sprintf(`
%[1]s

data "huaweicloud_availability_zones" "test" {}
data "huaweicloud_sdrs_domain" "test" {}

resource "huaweicloud_sdrs_protectiongroup" "test" {
  name                     = "%[2]s"
  source_availability_zone = data.huaweicloud_availability_zones.test.names[0]
  target_availability_zone = data.huaweicloud_availability_zones.test.names[1]
  domain_id                = data.huaweicloud_sdrs_domain.test.id
  source_vpc_id            = data.huaweicloud_vpc.test.id
  description              = "test description"
}

resource "huaweicloud_evs_volume" "test" {
  name              = "%[2]s"
  description       = "test volume for sdrs replication pair"
  availability_zone = data.huaweicloud_availability_zones.test.names[0]
  volume_type       = "SSD"
  size              = 100
}
`, common.TestVpc(name), name)
```

File: huaweicloud/services/sdrs/resource_huaweicloud_sdrs_replication_pair.go (hunk 7 above)
Comment (deer-hang) [reply huaweicloud/services/sdrs/resource_huaweicloud_sdrs_replication_pair.go:93]: This resource is implemented with reference to flexibleengine.
 Currently, change function `WaitForJobSuccess ` may cause error to flexibleengine.

**Comment by deer-hang** [comment]
```
make testacc TEST='./huaweicloud/services/acceptance/sdrs' TESTARGS='-run TestAccReplicationPair_basic'
==> Checking that code complies with gofmt requirements... 
TF_ACC=1 go test ./huaweicloud/services/acceptance/sdrs -v -run TestAccReplicationPair_basic -timeout 360m -parallel 4 
=== RUN   TestAccReplicationPair_basic 
=== PAUSE TestAccReplicationPair_basic
=== CONT  TestAccReplicationPair_basic
--- PASS: TestAccReplicationPair_basic (143.54s) 
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/sdrs      143.599s
```

File: huaweicloud/services/sdrs/resource_huaweicloud_sdrs_replication_pair.go (hunk 7 above)
Comment (chengxiangdong) [reply huaweicloud/services/sdrs/resource_huaweicloud_sdrs_replication_pair.go:93]: OK, I get it.

File: huaweicloud/services/sdrs/resource_huaweicloud_sdrs_replication_pair.go [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+	n, err := replications.Get(client, d.Id()).Extract()
+	if err != nil {
+		if errCode, ok := err.(golangsdk.ErrDefault400); ok {
+			if resp, pErr := common.ParseErrorMsg(errCode.Body); pErr == nil && resp.ErrorCode == "SDRS.1608" {
+				// `SDRS.1608` means replication pair not found
+				return common.CheckDeletedDiag(d, golangsdk.ErrDefault404{},
+					"error retrieving SDRS replication pair")
+			}
+		}
+		return diag.FromErr(err)
```
Comment (chengxiangdong) [inline huaweicloud/services/sdrs/resource_huaweicloud_sdrs_replication_pair.go:137]:
```suggestion
		return common.CheckDeletedDiag(d, err, "error retrieving SDRS replication pair")
```

File: huaweicloud/services/sdrs/resource_huaweicloud_sdrs_replication_pair.go [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+		return diag.Errorf("error creating SDRS client: %s", err)
+	}
+
+	n, err := replications.Get(client, d.Id()).Extract()
+	if err != nil {
+		if errCode, ok := err.(golangsdk.ErrDefault400); ok {
+			if resp, pErr := common.ParseErrorMsg(errCode.Body); pErr == nil && resp.ErrorCode == "SDRS.1608" {
+				// `SDRS.1608` means replication pair not found
+				return common.CheckDeletedDiag(d, golangsdk.ErrDefault404{},
+					"error retrieving SDRS replication pair")
```
Comment (chengxiangdong) [inline huaweicloud/services/sdrs/resource_huaweicloud_sdrs_replication_pair.go:134]:
```suggestion
				return diag.Errorf("invalid protected instance ID: %s", d.Id())
```

Comment (chengxiangdong) [reply huaweicloud/services/sdrs/resource_huaweicloud_sdrs_replication_pair.go:134]: Won't `SDRS.1608` be returned on create?

File: huaweicloud/services/sdrs/resource_huaweicloud_sdrs_replication_pair.go [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+	n, err := replications.Delete(client, d.Id(), deleteOpts).ExtractJobResponse()
+	if err != nil {
+		if errCode, ok := err.(golangsdk.ErrDefault400); ok {
+			if resp, pErr := common.ParseErrorMsg(errCode.Body); pErr == nil && resp.ErrorCode == "SDRS.1608" {
+				// `SDRS.1608` means replication pair not found
+				return common.CheckDeletedDiag(d, golangsdk.ErrDefault404{},
+					"error deleting SDRS replication pair")
+			}
+		}
+		return diag.FromErr(err)
```
Comment (chengxiangdong) [inline huaweicloud/services/sdrs/resource_huaweicloud_sdrs_replication_pair.go:204]:
```suggestion
		return common.CheckDeletedDiag(d, err, "error deletiing SDRS replication pair")
```

File: huaweicloud/services/sdrs/resource_huaweicloud_sdrs_replication_pair.go [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+	}
+
+	deleteTimeoutSec := int(d.Timeout(schema.TimeoutDelete).Seconds())
+	if err := replications.WaitForJobSuccess(client, deleteTimeoutSec, n.JobID); err != nil {
+		return diag.FromErr(err)
+	}
+	return nil
```
Comment (chengxiangdong) [inline huaweicloud/services/sdrs/resource_huaweicloud_sdrs_replication_pair.go:211]:
```suggestion
	d.SetId("")
	return nil
```

Comment (deer-hang) [reply huaweicloud/services/sdrs/resource_huaweicloud_sdrs_replication_pair.go:211]: The deletion method will automatically set the ID to empty internally.

Comment (chengxiangdong) [reply huaweicloud/services/sdrs/resource_huaweicloud_sdrs_replication_pair.go:211]: ok

**Comment by deer-hang** [comment]
```
make testacc TEST='./huaweicloud/services/acceptance/sdrs' TESTARGS='-run TestAccReplicationPair_basic'
==> Checking that code complies with gofmt requirements... 
TF_ACC=1 go test ./huaweicloud/services/acceptance/sdrs -v -run TestAccReplicationPair_basic -timeout 360m -parallel 4 
=== RUN   TestAccReplicationPair_basic 
=== PAUSE TestAccReplicationPair_basic
=== CONT  TestAccReplicationPair_basic
--- PASS: TestAccReplicationPair_basic (115.40s) 
PASS 
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/sdrs      115.465s
```
